pallets/providers/src/utils.rs

```diff
@@ -0,0 +1,185 @@
+use frame_support::pallet_prelude::DispatchResult;
+use frame_support::traits::Get;
+
+use crate::*;
+
+impl<T> Pallet<T>
+where
+    T: pallet::Config,
+{
+    pub fn do_msp_sign_up(who: &T::AccountId, msp_info: &MainStorageProvider<T>) -> DispatchResult {
+        // todo!()
+        <Msps<T>>::insert(&who, msp_info);
+        Ok(())
+    }
+
+    pub fn do_bsp_sign_up(
+        who: &T::AccountId,
+        bsp_info: BackupStorageProvider<T>,
+    ) -> DispatchResult {
+        // todo!()
+        <Bsps<T>>::insert(&who, bsp_info);
+        Ok(())
+    }
+
+    pub fn do_msp_sign_off(_who: &T::AccountId) -> DispatchResult {
+        todo!()
+    }
+
+    pub fn do_bsp_sign_off(_who: &T::AccountId) -> DispatchResult {
+        todo!()
+    }
+}
+
+impl<T: Config> From<MainStorageProvider<T>> for BackupStorageProvider<T> {
+    fn from(msp: MainStorageProvider<T>) -> Self {
+        BackupStorageProvider {
+            total_data: msp.total_data,
+            data_used: msp.data_used,
+            multiaddress: msp.multiaddress,
+            root: MerklePatriciaRoot::<T>::default(),
+        }
+    }
+}
+
+/// Implement the StorageProvidersInterface trait for the Storage Providers pallet.
+impl<T: pallet::Config> StorageProvidersInterface for pallet::Pallet<T> {
```

Comment:
^ based on my earlier comment, probably should add here `StorageProvidersInterface<T: Config>` and remove all associated types that are already defined in `Config` trait.